imap: Fallback to the LOGIN command when the AUTHENTICATE command fails

kop316 opened the pull request:
Hello!

This is a continuation of https://github.com/curl/curl/pull/8986

As per section 6.2.3 of RFC-9051 allow the fallback to a clear text LOGIN when a SASL based AUTHENTICATE fails.

I see two notes on the original PR:

https://github.com/curl/curl/pull/8986#issuecomment-1179588099

There's an option to prevent this without TLS, but what would break scripts: https://github.com/curl/curl/pull/8986#discussion_r894117162

The other note is: https://github.com/curl/curl/pull/8986#issuecomment-1190134446

The way I read the RFC, it looks like the same authentication should be used, but I may be wrong on that.

I am happy to attempt what I need to do to get this accepted. Thank you!

jay commented:
This appears to be a copy of @captain-caveman2k's proposal commit so he should be in the author field on this commit. This proposal was one of the times none of the collabs felt strongly either way so we didn't move on it. A bunch of CI jobs failed due test 849 which needs to be fixed, and that's where the PR went stale. I'd be less meh if this was only done over ssl, yes it would limit its usefulness but I don't know how that would break scripts.

/cc @emilengler

kop316 commented:
Hello!

> This appears to be a copy of @captain-caveman2k's proposal commit so he should be in the author field on this commit. 

That is correct, I changed the author of the commit.

> This proposal was one of the times none of the collabs felt strongly either way so we didn't move on it. A bunch of CI jobs failed due test 849 which needs to be fixed, and that's where the PR went stale. 

I can work through and see if I can fix them.

> I'd be less meh if this was only done over ssl, yes it would limit its usefulness but I don't know how that would break scripts.

I assume this is where the suggested change here: https://github.com/curl/curl/pull/8986#discussion_r894117162  Comes in? If so, I can try it and see if it fixes the issue I am facing.

> 
> /cc @emilengler

kop316 commented:
Hello!

It looks like as is, I was able to fix test 849. In trying the fix suggested here: https://github.com/curl/curl/pull/8986#discussion_r894117162

All of the IMAP tests work fine, except for the newly added one (`899`):
```
test 0800...[IMAP FETCH message]
--pd---e-v- OK (639 out of 1542, remaining: 20:06, took 1.366s, duration: 14:14)
test 0801...[IMAP FETCH message by MAILINDEX and SECTION]
--pd---e-v- OK (640 out of 1542, remaining: 20:05, took 1.330s, duration: 14:15)
test 0802...[IMAP SELECT UIDVALIDITY Success]
--pd---e-v- OK (641 out of 1542, remaining: 20:04, took 1.334s, duration: 14:16)
test 0803...[IMAP SELECT UIDVALIDITY Failure]
--p----e-v- OK (642 out of 1542, remaining: 20:02, took 1.295s, duration: 14:17)
test 0804...[IMAP doesn't perform SELECT if re-using the same mailbox]
--pd---e-v- OK (643 out of 1542, remaining: 20:01, took 1.394s, duration: 14:19)
test 0805...[IMAP APPEND message]
--p-u--e-v- OK (644 out of 1542, remaining: 20:00, took 1.289s, duration: 14:20)
test 0806...[IMAP LIST mailbox]
--pd---e-v- OK (645 out of 1542, remaining: 19:58, took 1.311s, duration: 14:21)
test 0807...[IMAP LSUB mailbox]
--pd---e-v- OK (646 out of 1542, remaining: 19:57, took 1.330s, duration: 14:23)
test 0808...[IMAP EXAMINE mailbox (CUSTOMREQUEST)]
--pd---e-v- OK (647 out of 1542, remaining: 19:56, took 1.351s, duration: 14:24)
test 0809...[IMAP mailbox STATUS (CUSTOMREQUEST)]
--pd---e-v- OK (648 out of 1542, remaining: 19:54, took 1.267s, duration: 14:25)
test 0810...[IMAP SEARCH for NEW messages]
--pd---e-v- OK (649 out of 1542, remaining: 19:53, took 1.377s, duration: 14:27)
test 0811...[IMAP CREATE mailbox (CUSTOMREQUEST)]
--p----e-v- OK (650 out of 1542, remaining: 19:51, took 1.253s, duration: 14:28)
test 0812...[IMAP DELETE mailbox (CUSTOMREQUEST)]
--p----e-v- OK (651 out of 1542, remaining: 19:50, took 1.309s, duration: 14:29)
test 0813...[IMAP RENAME mailbox (CUSTOMREQUEST)]
--p----e-v- OK (652 out of 1542, remaining: 19:49, took 1.330s, duration: 14:31)
test 0814...[IMAP CHECK mailbox (CUSTOMREQUEST)]
--p----e-v- OK (653 out of 1542, remaining: 19:47, took 1.331s, duration: 14:32)
test 0815...[IMAP STORE - delete message (CUSTOMREQUEST)]
--p----e-v- OK (654 out of 1542, remaining: 19:46, took 1.398s, duration: 14:33)
test 0816...[IMAP STORE - delete message with confirmation (CUSTOMREQUEST)]
--p----e-v- OK (655 out of 1542, remaining: 19:45, took 1.391s, duration: 14:35)
test 0817...[IMAP COPY message to mailbox (CUSTOMREQUEST)]
--p----e-v- OK (656 out of 1542, remaining: 19:43, took 1.248s, duration: 14:36)
test 0818...[IMAP NOOP (CUSTOMREQUEST)]
--pd---e-v- OK (657 out of 1542, remaining: 19:42, took 1.326s, duration: 14:37)
test 0819...[IMAP plain authentication]
--pd---e-v- OK (658 out of 1542, remaining: 19:41, took 1.363s, duration: 14:39)
test 0820...[IMAP login authentication]
--pd---e-v- OK (659 out of 1542, remaining: 19:39, took 1.378s, duration: 14:40)
test 0821...[IMAP CRAM-MD5 authentication]
--pd---e-v- OK (660 out of 1542, remaining: 19:38, took 1.385s, duration: 14:42)
test 0822...[IMAP NTLM authentication]
--pd---e-v- OK (661 out of 1542, remaining: 19:37, took 1.391s, duration: 14:43)
test 0824...[IMAP OAuth 2.0 (XOAUTH2) authentication]
--pd---e-v- OK (663 out of 1542, remaining: 19:33, took 1.381s, duration: 14:44)
test 0825...[IMAP plain authentication with initial response]
--pd---e-v- OK (664 out of 1542, remaining: 19:31, took 1.370s, duration: 14:46)
test 0826...[IMAP login authentication with initial response]
--pd---e-v- OK (665 out of 1542, remaining: 19:30, took 1.391s, duration: 14:47)
test 0827...[IMAP NTLM authentication with initial response]
--pd---e-v- OK (666 out of 1542, remaining: 19:29, took 1.449s, duration: 14:49)
test 0828...[IMAP OAuth 2.0 (XOAUTH2) authentication with initial response]
--pd---e-v- OK (667 out of 1542, remaining: 19:28, took 1.390s, duration: 14:50)
test 0829...[IMAP with URL-encoded CR LF in the URL]
-------e-v- OK (668 out of 1542, remaining: 19:26, took 1.232s, duration: 14:51)
test 0830...[IMAP CRAM-MD5 graceful cancellation]
--p----e-v- OK (669 out of 1542, remaining: 19:25, took 1.259s, duration: 14:52)
test 0831...[IMAP NTLM graceful cancellation]
--p----e-v- OK (670 out of 1542, remaining: 19:23, took 1.239s, duration: 14:54)
test 0833...[IMAP CRAM-MD5 authentication with SASL downgrade]
--pd---e-v- OK (672 out of 1542, remaining: 19:19, took 1.382s, duration: 14:55)
test 0834...[IMAP NTLM authentication with SASL downgrade]
--pd---e-v- OK (673 out of 1542, remaining: 19:18, took 1.427s, duration: 14:56)
test 0837...[IMAP external authentication]
--pd---e-v- OK (675 out of 1542, remaining: 19:13, took 1.391s, duration: 14:58)
test 0838...[IMAP external authentication without credentials]
--pd---e-v- OK (676 out of 1542, remaining: 19:12, took 1.361s, duration: 14:59)
test 0839...[IMAP external authentication with initial response]
--pd---e-v- OK (677 out of 1542, remaining: 19:11, took 1.370s, duration: 15:01)
test 0840...[IMAP external authentication with initial response without credentials]
--pd---e-v- OK (678 out of 1542, remaining: 19:10, took 1.398s, duration: 15:02)
test 0841...[IMAP custom request doesn't check continuation data]
--pd---e-v- OK (679 out of 1542, remaining: 19:08, took 1.334s, duration: 15:03)
test 0842...[IMAP OAuth 2.0 (OAUTHBEARER) authentication]
--pd---e-v- OK (680 out of 1542, remaining: 19:07, took 1.382s, duration: 15:05)
test 0843...[IMAP OAuth 2.0 (OAUTHBEARER) authentication with initial response]
--pd---e-v- OK (681 out of 1542, remaining: 19:06, took 1.366s, duration: 15:06)
test 0844...[IMAP OAuth 2.0 (OAUTHBEARER) failure as continuation]
--p----e-v- OK (682 out of 1542, remaining: 19:04, took 1.262s, duration: 15:07)
test 0845...[IMAP OAuth 2.0 (OAUTHBEARER) failure as continuation with initial response]
--p----e-v- OK (683 out of 1542, remaining: 19:03, took 1.312s, duration: 15:09)
test 0846...[IMAP PREAUTH response]
--pd---e-v- OK (684 out of 1542, remaining: 19:02, took 1.338s, duration: 15:10)
test 0847...[IMAP FETCH message]
--pd---e-v- OK (685 out of 1542, remaining: 19:00, took 1.402s, duration: 15:11)
test 0848...[IMAP plain authentication with alternative authorization identity]
--pd---e-v- OK (686 out of 1542, remaining: 18:59, took 1.372s, duration: 15:13)
test 0849...[IMAP plain authentication with alternative authorization identity (Not authorized)]
--p----e-v- OK (687 out of 1542, remaining: 18:58, took 1.235s, duration: 15:14)
test 0850...[POP3 RETR message]
--pd---e-v- OK (688 out of 1542, remaining: 18:58, took 2.365s, duration: 15:16)
test 0851...[POP3 LIST one message]
--p----e-v- OK (689 out of 1542, remaining: 18:56, took 1.189s, duration: 15:18)
test 0852...[POP3 LIST invalid message]
--p----e-v- OK (690 out of 1542, remaining: 18:55, took 1.187s, duration: 15:19)
test 0853...[POP3 LIST messages from *SLOW* server]
--pd---e-v- OK (691 out of 1542, remaining: 18:57, took 4.434s, duration: 15:23)
test 0854...[POP3 LIST no messages available]
--p----e-v- OK (692 out of 1542, remaining: 18:56, took 1.234s, duration: 15:24)
test 0855...[POP3 RETR invalid message]
--p----e-v- OK (693 out of 1542, remaining: 18:54, took 1.240s, duration: 15:26)
test 0856...[POP3 invalid login]
--p----e-v- OK (694 out of 1542, remaining: 18:53, took 1.296s, duration: 15:27)
test 0857...[POP3 RETR message with dot-prefixed line]
--pd---e-v- OK (695 out of 1542, remaining: 18:51, took 1.284s, duration: 15:28)
test 0858...[POP3 DELE message (CUSTOMREQUEST)]
--p----e-v- OK (696 out of 1542, remaining: 18:50, took 1.199s, duration: 15:29)
test 0859...[POP3 STAT (CUSTOMREQUEST)]
--p----e-v- OK (697 out of 1542, remaining: 18:48, took 1.200s, duration: 15:31)
test 0860...[POP3 NOOP (CUSTOMREQUEST)]
--p----e-v- OK (698 out of 1542, remaining: 18:47, took 1.200s, duration: 15:32)
test 0861...[POP3 UIDL (CUSTOMREQUEST)]
--pd---e-v- OK (699 out of 1542, remaining: 18:45, took 1.351s, duration: 15:33)
test 0862...[POP3 retrieve message header (CUSTOMREQUEST)]
--pd---e-v- OK (700 out of 1542, remaining: 18:44, took 1.326s, duration: 15:34)
test 0863...[POP3 RSET (CUSTOMREQUEST)]
--p----e-v- OK (701 out of 1542, remaining: 18:43, took 1.209s, duration: 15:36)
test 0864...[POP3 APOP authentication]
--pd---e-v- OK (702 out of 1542, remaining: 18:41, took 1.269s, duration: 15:37)
test 0865...[POP3 plain authentication]
--pd---e-v- OK (703 out of 1542, remaining: 18:40, took 1.307s, duration: 15:38)
test 0866...[POP3 login authentication]
--pd---e-v- OK (704 out of 1542, remaining: 18:38, took 1.284s, duration: 15:40)
test 0867...[POP3 CRAM-MD5 authentication]
--pd---e-v- OK (705 out of 1542, remaining: 18:37, took 1.300s, duration: 15:41)
test 0868...[POP3 NTLM authentication]
--pd---e-v- OK (706 out of 1542, remaining: 18:36, took 1.317s, duration: 15:42)
test 0870...[POP3 OAuth 2.0 (XOAUTH2) authentication]
--pd---e-v- OK (708 out of 1542, remaining: 18:31, took 1.318s, duration: 15:43)
test 0871...[POP3 plain authentication with initial response]
--pd---e-v- OK (709 out of 1542, remaining: 18:30, took 1.307s, duration: 15:45)
test 0872...[POP3 login authentication with initial response]
--pd---e-v- OK (710 out of 1542, remaining: 18:29, took 1.318s, duration: 15:46)
test 0873...[POP3 NTLM authentication with initial response]
--pd---e-v- OK (711 out of 1542, remaining: 18:27, took 1.330s, duration: 15:47)
test 0874...[POP3 OAuth 2.0 (XOAUTH2) authentication with initial response]
--pd---e-v- OK (712 out of 1542, remaining: 18:26, took 1.326s, duration: 15:49)
test 0875...[POP3 with URL-encoded CR LF in the URL]
-------e-v- OK (713 out of 1542, remaining: 18:25, took 1.189s, duration: 15:50)
test 0876...[POP3 CRAM-MD5 graceful cancellation]
--p----e-v- OK (714 out of 1542, remaining: 18:23, took 1.231s, duration: 15:51)
test 0877...[POP3 NTLM graceful cancellation]
--p----e-v- OK (715 out of 1542, remaining: 18:22, took 1.248s, duration: 15:52)
test 0879...[POP3 CRAM-MD5 authentication with SASL downgrade]
--pd---e-v- OK (717 out of 1542, remaining: 18:18, took 1.347s, duration: 15:54)
test 0880...[POP3 NTLM authentication with SASL downgrade]
--pd---e-v- OK (718 out of 1542, remaining: 18:16, took 1.328s, duration: 15:55)
test 0883...[POP3 external authentication]
--pd---e-v- OK (720 out of 1542, remaining: 18:12, took 1.306s, duration: 15:56)
test 0884...[POP3 external authentication without credentials]
--pd---e-v- OK (721 out of 1542, remaining: 18:11, took 1.368s, duration: 15:58)
test 0885...[POP3 external authentication with initial response]
--pd---e-v- OK (722 out of 1542, remaining: 18:09, took 1.336s, duration: 15:59)
test 0886...[POP3 external authentication with initial response without credentials]
--pd---e-v- OK (723 out of 1542, remaining: 18:08, took 1.324s, duration: 16:00)
test 0887...[POP3 OAuth 2.0 (OAUTHBEARER) authentication]
--pd---e-v- OK (724 out of 1542, remaining: 18:07, took 1.293s, duration: 16:02)
test 0888...[POP3 OAuth 2.0 (OAUTHBEARER) authentication with initial response]
--pd---e-v- OK (725 out of 1542, remaining: 18:05, took 1.303s, duration: 16:03)
test 0889...[POP3 OAuth 2.0 (OAUTHBEARER) failure as continuation]
--p----e-v- OK (726 out of 1542, remaining: 18:06, took 2.791s, duration: 16:06)
test 0890...[POP3 OAuth 2.0 (OAUTHBEARER) failure as continuation with initial response]
--p----e-v- OK (727 out of 1542, remaining: 18:04, took 1.227s, duration: 16:07)
test 0891...[POP3 with short authentication response]
--p----e-v- OK (728 out of 1542, remaining: 18:03, took 1.246s, duration: 16:08)
test 0892...[POP3 plain authentication with alternative authorization identity]
--pd---e-v- OK (729 out of 1542, remaining: 18:01, took 1.315s, duration: 16:10)
test 0893...[POP3 plain authentication with alternative authorization identity (Not authorized)]
--p----e-v- OK (730 out of 1542, remaining: 18:00, took 1.222s, duration: 16:11)
test 0894...[POP3 with CR in username]
-------e-v- OK (731 out of 1542, remaining: 17:58, took 1.090s, duration: 16:12)
test 0895...[IMAP with --login-options 'AUTH=*']
--pd---e-v- OK (732 out of 1542, remaining: 17:57, took 1.312s, duration: 16:13)
test 0896...[IMAP with --login-options 'AUTH=dummy' (failing)]
-------e-v- OK (733 out of 1542, remaining: 17:55, took 1.169s, duration: 16:14)
test 0897...[IMAP and envelope meta data after body transfer]
--pd--oe-v- OK (734 out of 1542, remaining: 17:54, took 1.329s, duration: 16:16)
test 0898...[HTTP with custom auth and cookies redirected to HTTP on a diff port]
--pd---e-v- OK (735 out of 1542, remaining: 17:53, took 1.238s, duration: 16:17)
test 0899...[IMAPS plain authentication with fallback to clear text]

 899: protocol FAILED:
--- log/check-expected	2022-12-05 08:39:24.659860594 -0500
+++ log/check-generated	2022-12-05 08:39:24.659860594 -0500
@@ -1,6 +1,4 @@
-A001 CAPABILITY[CR][LF]
-A002 AUTHENTICATE PLAIN AHVzZXIAc2VjcmV0[CR][LF]
-A003 LOGIN user secret[CR][LF]
-A004 SELECT 899[CR][LF]
-A005 FETCH 1 BODY[][CR][LF]
-A006 LOGOUT[CR][LF]
+%16%03%01%02%00%01%00%01%fc%03%03%94%fbT%9bf%d3J%88GI%12%dbi%b4%eb%a5n%8f&%c4%f5E%fd%9e>.T%aaE%db%a5%b2 &%a7%12(%03%e0%ad:l%03%f7%cb[n_(%e7%d0%bdA%c0%9a%1e%95%fc#%bd%87Id%c4%10%00>%13%02%13%03%13%01%c0,%c00%00%9f%cc%a9%cc%a8%cc%aa%c0+%c0/%00%9e%c0$%c0(%00k%c0#%c0'%00g%c0[LF]
+%c0%14%009%c0%09%c0%13%003%00%9d%00%9c%00=%00<%005%00/%00%ff%01%00%01u%00%0b%00%04%03%00%01%02%00[LF]
+%00%16%00%14%00%1d%00%17%00%1e%00%19%00%18%01%00%01%01%01%02%01%03%01%04%00%16%00%00%00%17%00%00%001%00%00%00[CR]%00*%00(%04%03%05%03%06%03%08%07%08%08%08%09%08[LF]
+%08%0b%08%04%08%05%08%06%04%01%05%01%06%01%03%03%03%01%03%02%04%02%05%02%06%02%00+%00%09%08%03%04%03%03%03%02%03%01%00-%00%02%01%01%003%00&%00$%00%1d%00 ;+%dc-%cf%008n%0c%89%01%a4%92Z%ed%01%ce%bc%c8%89%e0%c5%e9%1a%b7%b4%a1%ab8$%12h%00%15%00%d8%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00
== Contents of files in the log/ dir after test 899
=== Start of file check-expected
 A001 CAPABILITY[CR][LF]
 A002 AUTHENTICATE PLAIN AHVzZXIAc2VjcmV0[CR][LF]
 A003 LOGIN user secret[CR][LF]
 A004 SELECT 899[CR][LF]
 A005 FETCH 1 BODY[][CR][LF]
 A006 LOGOUT[CR][LF]
=== End of file check-expected
=== Start of file check-generated
 %16%03%01%02%00%01%00%01%fc%03%03%94%fbT%9bf%d3J%88GI%12%dbi%b4%eb%a5n%8f&%c4%f5E%fd%9e>.T%aaE%db%a5%b2 &%a7%12(%03%e0%ad:l%03%f7%cb[n_(%e7%d0%bdA%c0%9a%1e%95%fc#%bd%87Id%c4%10%00>%13%02%13%03%13%01%c0,%c00%00%9f%cc%a9%cc%a8%cc%aa%c0+%c0/%00%9e%c0$%c0(%00k%c0#%c0'%00g%c0[LF]
 %c0%14%009%c0%09%c0%13%003%00%9d%00%9c%00=%00<%005%00/%00%ff%01%00%01u%00%0b%00%04%03%00%01%02%00[LF]
 %00%16%00%14%00%1d%00%17%00%1e%00%19%00%18%01%00%01%01%01%02%01%03%01%04%00%16%00%00%00%17%00%00%001%00%00%00[CR]%00*%00(%04%03%05%03%06%03%08%07%08%08%08%09%08[LF]
 %08%0b%08%04%08%05%08%06%04%01%05%01%06%01%03%03%03%01%03%02%04%02%05%02%06%02%00+%00%09%08%03%04%03%03%03%02%03%01%00-%00%02%01%01%003%00&%00$%00%1d%00 ;+%dc-%cf%008n%0c%89%01%a4%92Z%ed%01%ce%bc%c8%89%e0%c5%e9%1a%b7%b4%a1%ab8$%12h%00%15%00%d8%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00
=== End of file check-generated
=== Start of file commands.log
 ../libtool --mode=execute /usr/bin/valgrind --tool=memcheck --quiet --leak-check=yes --suppressions=./valgrind.supp --num-callers=16 --log-file=log/valgrind899 ../src/curl --output log/curl899.out  --include --trace-ascii log/trace899 --trace-time 'imaps://127.0.0.1:43085/899/;MAILINDEX=1' -u user:secret >log/stdout899 2>log/stderr899
=== End of file commands.log
=== Start of file ftpserver.cmd
 AUTH PLAIN
 CAPA SASL-IR
 REPLY AUTHENTICATE A002 NO AUTHENTICATE failed
 REPLY LOGIN A003 OK LOGIN completed
 Testnum 899
=== End of file ftpserver.cmd
=== Start of file imap_server.log
 08:39:22.745658 ====> Client connect
 08:39:22.745736 FTPD: Getting commands from log/ftpserver.cmd
 08:39:22.745764 FTPD: instructed to support AUTHENTICATION command
 08:39:22.745781 FTPD: instructed to support CAPABILITY command
 08:39:22.745838 FTPD: set custom reply for AUTHENTICATE command
 08:39:22.745870 FTPD: set custom reply for LOGIN command
 08:39:22.745888 FTPD: run test case number: 899
 08:39:22.745950 > "        _   _ ____  _     [CR][LF]    ___| | | |  _ \| |    [CR][LF]   / __| | | | |_) | |    [CR][LF]  | (__| |_| |  _ {| |___ [CR][LF]   \___|\___/|_| \_\_____|[CR][LF]* OK curl IMAP server ready to serve[CR][LF]"
 08:39:24.564741 MAIN sockfilt said DISC
 08:39:24.564787 ====> Client disconnected
 08:39:24.564833 Awaiting input
=== End of file imap_server.log
=== Start of file imap_sockfilt.log
 08:39:22.665338 ====> Client connect
 08:39:22.665706 Received DATA (on stdin)
 08:39:22.665725 > 178 bytes data, server => client
 08:39:22.665740 '        _   _ ____  _     \r\n    ___| | | |  _ \| |    \r\n '
 08:39:22.665748 '  / __| | | | |_) | |    \r\n  | (__| |_| |  _ {| |___ \r\n  '
 08:39:22.665755 ' \___|\___/|_| \_\_____|\r\n* OK curl IMAP server ready to se'
 08:39:22.665761 'rve\r\n'
 08:39:24.439134 < 517 bytes data, client => server
 08:39:24.439175 '.............T.f.J.GI..i...n.&..E..>.T.E... &..(...:l...[n_(.'
 08:39:24.439188 '..A.....#..Id...>.......,.0.........+./...$.(.k.#.'.g.\n...9.'
 08:39:24.439200 '	...3.....=.<.5./.....u.
                                               .......\n...........................'
 08:39:24.439212 '......1...\r.*.(...........	.\n.
                                                    .........................+.	'
 08:39:24.439243 '..........-.....3.&.$... ;+.-..8n
 ....Z..............8$.h....'
 08:39:24.439255 '.............................................................'
 08:39:24.439268 '.............................................................'
 08:39:24.439279 '.............................................................'
 08:39:24.439285 '.................................'
 08:39:24.484438 ====> Client disconnect
=== End of file imap_sockfilt.log
=== Start of file server.input
 ���T�f�J�GI�i��n�&��E��>.T�Eۥ� &�(�:l��[n_(�нA����#��Id�>�,�0�̨̩̪�+�/��$�(k�#�'g�
 �9�	�3��=<5/�u

 1
 *(
 
+      -3&$ ;+�-�8n
 ���Z�μȉ���
 ����8$h�
=== End of file server.input
=== Start of file stderr899
   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                  Dload  Upload   Total   Spent    Left  Speed
 
   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
   0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
 curl: (35) OpenSSL/3.0.7: error:0A00010B:SSL routines::wrong version number
=== End of file stderr899
=== Start of file trace899
 08:39:22.643319 == Info:   Trying 127.0.0.1:43085...
 08:39:22.889476 == Info: Connected to 127.0.0.1 (127.0.0.1) port 43085 (#0)
 08:39:24.264490 == Info:  CAfile: /etc/ssl/certs/ca-certificates.crt
 08:39:24.264629 == Info:  CApath: none
 08:39:24.428686 == Info: [CONN-0-0][CF-SSL] TLSv1.0 (OUT), TLS header, Certificate Status (22):
 08:39:24.428883 => Send SSL data, 5 bytes (0x5)
 0000: .....
 08:39:24.436548 == Info: [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Client hello (1):
 08:39:24.436595 => Send SSL data, 512 bytes (0x200)
 0000: ........T.f.J.GI..i...n.&..E..>.T.E... &..(...:l...[n_(...A.....
 0040: #..Id...>.......,.0.........+./...$.(.k.#.'.g.....9.....3.....=.
 0080: <.5./.....u...........................................1.....*.(.
 00c0: ........................................+............-.....3.&.$
 0100: ... ;+.-..8n.....Z..............8$.h............................
 0140: ................................................................
 0180: ................................................................
 01c0: ................................................................
 08:39:24.445572 == Info: [CONN-0-0][CF-SSL] (2020) (IN), , Unknown (32):
 08:39:24.445620 <= Recv SSL data, 5 bytes (0x5)
 0000:      
 08:39:24.453318 == Info: OpenSSL/3.0.7: error:0A00010B:SSL routines::wrong version number
 08:39:24.464627 == Info: Closing connection 0
=== End of file trace899
```
It looks like there is an issue using openssl with the ftpserver.pl?

The patch applied is here:
```
diff --git a/lib/imap.c b/lib/imap.c
index 95f9a1507..ee3b50243 100644
--- a/lib/imap.c
+++ b/lib/imap.c
@@ -1030,7 +1030,8 @@ static CURLcode imap_state_auth_resp(struct Curl_easy *data,
     default:
       break;
     }
-  else if((!imapc->login_disabled) && (imapc->preftype & IMAP_TYPE_CLEARTEXT))
+  else if(!imapc->login_disabled && imapc->ssldone &&
+          (imapc->preftype & IMAP_TYPE_CLEARTEXT))
     /* Perform clear text authentication on failure as per RFC-9051 */
     result = imap_perform_login(data, conn);
 
diff --git a/tests/data/test849 b/tests/data/test849
index e4c680235..20fca6b94 100644
--- a/tests/data/test849
+++ b/tests/data/test849
@@ -15,7 +15,6 @@ RFC4616
 AUTH PLAIN
 REPLY AUTHENTICATE +
 REPLY dXJzZWwAa3VydAB4aXBqM3BsbXE= A002 NO Not authorized
-REPLY LOGIN A003 NO LOGIN failed
 </servercmd>
 </reply>
 
@@ -47,7 +46,6 @@ IMAP plain authentication with alternative authorization identity (Not authorize
 A001 CAPABILITY
 A002 AUTHENTICATE PLAIN
 dXJzZWwAa3VydAB4aXBqM3BsbXE=
-A003 LOGIN kurt xipj3plmq
 </protocol>
 </verify>
 </testcase>
diff --git a/tests/data/test899 b/tests/data/test899
index d50e102eb..2d04ba96d 100644
--- a/tests/data/test899
+++ b/tests/data/test899
@@ -41,7 +41,7 @@ imap
 IMAP plain authentication with fallback to clear text
  </name>
  <command>
-'imap://%HOSTIP:%IMAPPORT/%TESTNUMBER/;MAILINDEX=1' -u user:secret
+'imaps://%HOSTIP:%IMAPPORT/%TESTNUMBER/;MAILINDEX=1' -u user:secret
 </command>
 </client>
 ```

vszakats commented:
Just curious, how can this feature be disabled for those who want to avoid falling back to unsecure cleartext auth? (and thus avoid a downgrade attach, unless I'm missing something)

kop316 commented:
> Just curious, how can this feature be disabled for those who want to avoid falling back to unsecure cleartext auth? (and thus avoid a downgrade attack, unless I'm missing something)

My understanding is to use `CURLOPT_USE_SSL` with the option `CURLUSESSL_CONTROL` or `CURLUSESSL_ALL` . This forces STARTTLS if the protocol is IMAP vs. IMAPS.

kop316 commented:
> Perhaps the solution is to have a check in `imap_perform_login` to see if SSL is enabled and disallow over plaintext?

So several IMAP test benches are dependent on `imap_perform_login()`, so changing https://github.com/curl/curl/blob/master/lib/imap.c#L628 and https://github.com/curl/curl/blob/master/lib/imap.c#L1023 would require fixing a bunch of test scripts.

As noted here, forcing SSL would only break the added script (`test899`): https://github.com/curl/curl/pull/10027#issuecomment-1337394312 , so maybe I should just delete that script and add the SSL?

I was also looking into this, according to `RFC3501` (Original or Rev1), Section 6.3.2: A server must advertise LOGINDISABLED if it has no method to protect against password snooping. So in theory, curl would never use the `imap_perform_login()` (but I imagine there are misconfigured servers out that would allow this anyways....)

To reiterate, I am fine with either solution (leave as is or only allow it with SSL and delete `test899`).

vszakats commented:
As a non-expert IMAP user, my preference goes with the option that avoids leaking credentials, either by accident, due to misconfigured client or server, or due to a malicious 3rd-party. Tying this fallback to the presence of a connected/verified SSL endpoint (by default or always) is probably enough.

kop316 commented:
> As a non-expert IMAP user, my preference goes with the option that avoids leaking credentials, either by accident, due to misconfigured client or server, or due to a malicious 3rd-party. Tying this fallback to the presence of a connected/verified SSL endpoint (by default or always) is probably enough.

That works for me! I updated the MR to only allow for the login fallback as indicated in https://github.com/curl/curl/pull/8986#discussion_r894117162 . I also removed `test899` since the server does not have  SSL capabilities.

jay commented:
The test could be modified to use imaps could it not?

kop316 commented:
> The test could be modified to use imaps could it not?

I attempted that earlier in https://github.com/curl/curl/pull/10027#issuecomment-1337394312

From what I can tell, ftpserver.pl does not implement openssl at all, and current, no test exercises imaps (all imap tests are done without SSL). I looked to try and see how ftps is tested, but it looks like only the control channel is unecrypted in all ftps tests, and I was unable to figure out how the control channel was using openssl.

This is the result of the test when modified for imaps:
```
test 0899...[IMAPS plain authentication with fallback to clear text]

 899: protocol FAILED:
--- log/check-expected	2022-12-05 08:39:24.659860594 -0500
+++ log/check-generated	2022-12-05 08:39:24.659860594 -0500
@@ -1,6 +1,4 @@
-A001 CAPABILITY[CR][LF]
-A002 AUTHENTICATE PLAIN AHVzZXIAc2VjcmV0[CR][LF]
-A003 LOGIN user secret[CR][LF]
-A004 SELECT 899[CR][LF]
-A005 FETCH 1 BODY[][CR][LF]
-A006 LOGOUT[CR][LF]
+%16%03%01%02%00%01%00%01%fc%03%03%94%fbT%9bf%d3J%88GI%12%dbi%b4%eb%a5n%8f&%c4%f5E%fd%9e>.T%aaE%db%a5%b2 &%a7%12(%03%e0%ad:l%03%f7%cb[n_(%e7%d0%bdA%c0%9a%1e%95%fc#%bd%87Id%c4%10%00>%13%02%13%03%13%01%c0,%c00%00%9f%cc%a9%cc%a8%cc%aa%c0+%c0/%00%9e%c0$%c0(%00k%c0#%c0'%00g%c0[LF]
+%c0%14%009%c0%09%c0%13%003%00%9d%00%9c%00=%00<%005%00/%00%ff%01%00%01u%00%0b%00%04%03%00%01%02%00[LF]
+%00%16%00%14%00%1d%00%17%00%1e%00%19%00%18%01%00%01%01%01%02%01%03%01%04%00%16%00%00%00%17%00%00%001%00%00%00[CR]%00*%00(%04%03%05%03%06%03%08%07%08%08%08%09%08[LF]
+%08%0b%08%04%08%05%08%06%04%01%05%01%06%01%03%03%03%01%03%02%04%02%05%02%06%02%00+%00%09%08%03%04%03%03%03%02%03%01%00-%00%02%01%01%003%00&%00$%00%1d%00 ;+%dc-%cf%008n%0c%89%01%a4%92Z%ed%01%ce%bc%c8%89%e0%c5%e9%1a%b7%b4%a1%ab8$%12h%00%15%00%d8%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00
== Contents of files in the log/ dir after test 899
=== Start of file check-expected
 A001 CAPABILITY[CR][LF]
 A002 AUTHENTICATE PLAIN AHVzZXIAc2VjcmV0[CR][LF]
 A003 LOGIN user secret[CR][LF]
 A004 SELECT 899[CR][LF]
 A005 FETCH 1 BODY[][CR][LF]
 A006 LOGOUT[CR][LF]
=== End of file check-expected
=== Start of file check-generated
 %16%03%01%02%00%01%00%01%fc%03%03%94%fbT%9bf%d3J%88GI%12%dbi%b4%eb%a5n%8f&%c4%f5E%fd%9e>.T%aaE%db%a5%b2 &%a7%12(%03%e0%ad:l%03%f7%cb[n_(%e7%d0%bdA%c0%9a%1e%95%fc#%bd%87Id%c4%10%00>%13%02%13%03%13%01%c0,%c00%00%9f%cc%a9%cc%a8%cc%aa%c0+%c0/%00%9e%c0$%c0(%00k%c0#%c0'%00g%c0[LF]
 %c0%14%009%c0%09%c0%13%003%00%9d%00%9c%00=%00<%005%00/%00%ff%01%00%01u%00%0b%00%04%03%00%01%02%00[LF]
 %00%16%00%14%00%1d%00%17%00%1e%00%19%00%18%01%00%01%01%01%02%01%03%01%04%00%16%00%00%00%17%00%00%001%00%00%00[CR]%00*%00(%04%03%05%03%06%03%08%07%08%08%08%09%08[LF]
 %08%0b%08%04%08%05%08%06%04%01%05%01%06%01%03%03%03%01%03%02%04%02%05%02%06%02%00+%00%09%08%03%04%03%03%03%02%03%01%00-%00%02%01%01%003%00&%00$%00%1d%00 ;+%dc-%cf%008n%0c%89%01%a4%92Z%ed%01%ce%bc%c8%89%e0%c5%e9%1a%b7%b4%a1%ab8$%12h%00%15%00%d8%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00
=== End of file check-generated
=== Start of file commands.log
 ../libtool --mode=execute /usr/bin/valgrind --tool=memcheck --quiet --leak-check=yes --suppressions=./valgrind.supp --num-callers=16 --log-file=log/valgrind899 ../src/curl --output log/curl899.out  --include --trace-ascii log/trace899 --trace-time 'imaps://127.0.0.1:43085/899/;MAILINDEX=1' -u user:secret >log/stdout899 2>log/stderr899
=== End of file commands.log
=== Start of file ftpserver.cmd
 AUTH PLAIN
 CAPA SASL-IR
 REPLY AUTHENTICATE A002 NO AUTHENTICATE failed
 REPLY LOGIN A003 OK LOGIN completed
 Testnum 899
=== End of file ftpserver.cmd
=== Start of file imap_server.log
 08:39:22.745658 ====> Client connect
 08:39:22.745736 FTPD: Getting commands from log/ftpserver.cmd
 08:39:22.745764 FTPD: instructed to support AUTHENTICATION command
 08:39:22.745781 FTPD: instructed to support CAPABILITY command
 08:39:22.745838 FTPD: set custom reply for AUTHENTICATE command
 08:39:22.745870 FTPD: set custom reply for LOGIN command
 08:39:22.745888 FTPD: run test case number: 899
 08:39:22.745950 > "        _   _ ____  _     [CR][LF]    ___| | | |  _ \| |    [CR][LF]   / __| | | | |_) | |    [CR][LF]  | (__| |_| |  _ {| |___ [CR][LF]   \___|\___/|_| \_\_____|[CR][LF]* OK curl IMAP server ready to serve[CR][LF]"
 08:39:24.564741 MAIN sockfilt said DISC
 08:39:24.564787 ====> Client disconnected
 08:39:24.564833 Awaiting input
=== End of file imap_server.log
=== Start of file imap_sockfilt.log
 08:39:22.665338 ====> Client connect
 08:39:22.665706 Received DATA (on stdin)
 08:39:22.665725 > 178 bytes data, server => client
 08:39:22.665740 '        _   _ ____  _     \r\n    ___| | | |  _ \| |    \r\n '
 08:39:22.665748 '  / __| | | | |_) | |    \r\n  | (__| |_| |  _ {| |___ \r\n  '
 08:39:22.665755 ' \___|\___/|_| \_\_____|\r\n* OK curl IMAP server ready to se'
 08:39:22.665761 'rve\r\n'
 08:39:24.439134 < 517 bytes data, client => server
 08:39:24.439175 '.............T.f.J.GI..i...n.&..E..>.T.E... &..(...:l...[n_(.'
 08:39:24.439188 '..A.....#..Id...>.......,.0.........+./...$.(.k.#.'.g.\n...9.'
 08:39:24.439200 '	...3.....=.<.5./.....u.
                                               .......\n...........................'
 08:39:24.439212 '......1...\r.*.(...........	.\n.
                                                    .........................+.	'
 08:39:24.439243 '..........-.....3.&.$... ;+.-..8n
 ....Z..............8$.h....'
 08:39:24.439255 '.............................................................'
 08:39:24.439268 '.............................................................'
 08:39:24.439279 '.............................................................'
 08:39:24.439285 '.................................'
 08:39:24.484438 ====> Client disconnect
=== End of file imap_sockfilt.log
=== Start of file server.input
 ���T�f�J�GI�i��n�&��E��>.T�Eۥ� &�(�:l��[n_(�нA����#��Id�>�,�0�̨̩̪�+�/��$�(k�#�'g�
 �9�	�3��=<5/�u

 1
 *(
 
+      -3&$ ;+�-�8n
 ���Z�μȉ���
 ����8$h�
=== End of file server.input
=== Start of file stderr899
   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                  Dload  Upload   Total   Spent    Left  Speed
 
   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
   0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
 curl: (35) OpenSSL/3.0.7: error:0A00010B:SSL routines::wrong version number
=== End of file stderr899
=== Start of file trace899
 08:39:22.643319 == Info:   Trying 127.0.0.1:43085...
 08:39:22.889476 == Info: Connected to 127.0.0.1 (127.0.0.1) port 43085 (#0)
 08:39:24.264490 == Info:  CAfile: /etc/ssl/certs/ca-certificates.crt
 08:39:24.264629 == Info:  CApath: none
 08:39:24.428686 == Info: [CONN-0-0][CF-SSL] TLSv1.0 (OUT), TLS header, Certificate Status (22):
 08:39:24.428883 => Send SSL data, 5 bytes (0x5)
 0000: .....
 08:39:24.436548 == Info: [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Client hello (1):
 08:39:24.436595 => Send SSL data, 512 bytes (0x200)
 0000: ........T.f.J.GI..i...n.&..E..>.T.E... &..(...:l...[n_(...A.....
 0040: #..Id...>.......,.0.........+./...$.(.k.#.'.g.....9.....3.....=.
 0080: <.5./.....u...........................................1.....*.(.
 00c0: ........................................+............-.....3.&.$
 0100: ... ;+.-..8n.....Z..............8$.h............................
 0140: ................................................................
 0180: ................................................................
 01c0: ................................................................
 08:39:24.445572 == Info: [CONN-0-0][CF-SSL] (2020) (IN), , Unknown (32):
 08:39:24.445620 <= Recv SSL data, 5 bytes (0x5)
 0000:      
 08:39:24.453318 == Info: OpenSSL/3.0.7: error:0A00010B:SSL routines::wrong version number
 08:39:24.464627 == Info: Closing connection 0
=== End of file trace899
```
This is how I modified the test:
```
diff --git a/tests/data/test849 b/tests/data/test849
index e4c680235..20fca6b94 100644
--- a/tests/data/test849
+++ b/tests/data/test849
@@ -15,7 +15,6 @@ RFC4616
 AUTH PLAIN
 REPLY AUTHENTICATE +
 REPLY dXJzZWwAa3VydAB4aXBqM3BsbXE= A002 NO Not authorized
-REPLY LOGIN A003 NO LOGIN failed
 </servercmd>
 </reply>
 
@@ -47,7 +46,6 @@ IMAP plain authentication with alternative authorization identity (Not authorize
 A001 CAPABILITY
 A002 AUTHENTICATE PLAIN
 dXJzZWwAa3VydAB4aXBqM3BsbXE=
-A003 LOGIN kurt xipj3plmq
 </protocol>
 </verify>
 </testcase>
diff --git a/tests/data/test899 b/tests/data/test899
index d50e102eb..2d04ba96d 100644
--- a/tests/data/test899
+++ b/tests/data/test899
@@ -41,7 +41,7 @@ imap
 IMAP plain authentication with fallback to clear text
  </name>
  <command>
-'imap://%HOSTIP:%IMAPPORT/%TESTNUMBER/;MAILINDEX=1' -u user:secret
+'imaps://%HOSTIP:%IMAPPORT/%TESTNUMBER/;MAILINDEX=1' -u user:secret
 </command>
 </client>
 ```

jay commented:
Thanks, I see that you did attempt that. It looks like runtests currently does not support imaps for server type. I will look into this. I would like to restore the proposed test if it's feasible. Also, I would like to hear from @emilengler how restricting the fallback to SSL would break scripts.

emilengler commented:
> Thanks, I see that you did attempt that. It looks like runtests currently does not support imaps for server type. I will look into this. I would like to restore the proposed test if it's feasible. Also, I would like to hear from @emilengler how restricting the fallback to SSL would break scripts.

The only circumstance I can imagine, under which scripts would break, would be scripts explicitly trying to use SASL for authentication. This may be unit tests for IMAP servers, analysis tools, etc. It is a niche thing, but I would suggest to at least and a line to the log, in which it informs about the fallback.

kop316 commented:
> Thanks, I see that you did attempt that. It looks like runtests currently does not support imaps for server type. I will look into this. I would like to restore the proposed test if it's feasible. 

Sounds good to me, thank you for looking into it! If we can get the imaps server type running, I agree, there should be a test for it.

kop316 commented on lib/imap.c:
```diff
@@ -1029,6 +1030,10 @@ static CURLcode imap_state_auth_resp(struct Curl_easy *data,
       break;
     }
+  else if(!imapc->login_disabled && imapc->ssldone &&
+          (imapc->preftype & IMAP_TYPE_CLEARTEXT))
```
> It is a niche thing, but I would suggest to at least and a line to the log, in which it informs about the fallback.

It sounds like you would like:

```
  else if(!imapc->login_disabled && imapc->ssldone &&
          (imapc->preftype & IMAP_TYPE_CLEARTEXT)) {
    /* Perform clear text authentication on failure as per RFC-9051 */
    infof(data, "Falling back to IMAP LOGIN");
    result = imap_perform_login(data, conn);
  }
```

If so, I can add that.

kop316 replied:
Ok, I added that line.

kop316 commented:
> The only circumstance I can imagine, under which scripts would break, would be scripts explicitly trying to use SASL for authentication. This may be unit tests for IMAP servers, analysis tools, etc.

If I understand this correctly, you mean if someone uses e.g. `curl -v --login-options AUTH=$FOO imaps://$USER:$PW@$HOST/ `

The code is never reached as `(imapc->preftype & IMAP_TYPE_CLEARTEXT)` returns false. In `imap_parse_url_options()`, the `IMAP_TYPE_CLEARTEXT` bit is not set when the option `AUTH=$FOO` is applied.

https://github.com/curl/curl/blob/master/lib/imap.c#L1919

Here, I explicitly do that with the PR applied:

```
chris@SpaceballsTheX2100:~/Downloads/curl$ src/curl -v --login-options AUTH=DIGEST-MD5 imaps://$REDACTED_USER:$REDACTED_PW@vvm.mstore.msg.t-mobile.com/
*   Trying 208.54.156.116:993...
* Connected to vvm.mstore.msg.t-mobile.com (208.54.156.116) port 993 (#0)
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* [CONN-0-0][CF-SSL] TLSv1.0 (OUT), TLS header, Certificate Status (22):
* [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Client hello (1):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Certificate Status (22):
* [CONN-0-0][CF-SSL] TLSv1.3 (IN), TLS handshake, Server hello (2):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Certificate Status (22):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS handshake, Certificate (11):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Certificate Status (22):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS handshake, Server finished (14):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS header, Certificate Status (22):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS header, Finished (20):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS header, Certificate Status (22):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS handshake, Finished (20):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Finished (20):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Certificate Status (22):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / AES256-GCM-SHA384
* Server certificate:
*  subject: C=US; ST=Washington; L=Bellevue; O=T-Mobile USA, Inc.; CN=vvm.mstore.msg.t-mobile.com
*  start date: Jun  7 20:06:33 2022 GMT
*  expire date: Jun  7 20:06:33 2023 GMT
*  subjectAltName: host "vvm.mstore.msg.t-mobile.com" matched cert's "vvm.mstore.msg.t-mobile.com"
*  issuer: C=US; O=Entrust, Inc.; OU=See www.entrust.net/legal-terms; OU=(c) 2012 Entrust, Inc. - for authorized use only; CN=Entrust Certification Authority - L1K
*  SSL certificate verify ok.
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Supplemental data (23):
< * OK [CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS] mStore IMAP4rev1 Server is ready.
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS header, Supplemental data (23):
> A001 CAPABILITY
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Supplemental data (23):
< * CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Supplemental data (23):
< A001 OK CAPABILITY completed.
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS header, Supplemental data (23):
> A002 AUTHENTICATE DIGEST-MD5
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Supplemental data (23):
< + $REDACTED_MD5
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS header, Supplemental data (23):
> $REDACTED_MD5
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Supplemental data (23):
< A002 NO application error
* Closing connection 0
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS header, Unknown (21):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS alert, close notify (256):
curl: (67) Login denied
```

jay commented:
> It looks like runtests currently does not support imaps for server type. I will look into this. I would like to restore the proposed test if it's feasible.

@monnerat has proposed #10077 which adds imaps support to tests, so after that lands you can restore the test.

monnerat commented:
Please, see comment https://github.com/curl/curl/pull/8986#issuecomment-1190134446

I'm still convinced the authentication mechanism must be part of the connection cache selection filtering if we commit this or else we would introduce a security problem.

kop316 commented:
> Please, see comment [#8986 (comment)](https://github.com/curl/curl/pull/8986#issuecomment-1190134446)
> 
> I'm still convinced the authentication mechanism must be part of the connection cache selection filtering if we commit this or else we would introduce a security problem.

I am trying to understand this part, as when I looked through the IMAP tests, I didn't see any with multiple user/PW pairs. Are you trying to say depending on the AUTHENTICATE command, (e.g. PLAIN, DIGEST-MD5, etc.), we may not want to use this fallback at all? Or am I misunderstanding the intent?

My understanding of how this works in in line with: https://github.com/curl/curl/pull/8986#issuecomment-1193198363

Essentially, AUTHENTICATE is broken for some reason (in the case I am trying to solve, the server is badly misconfigured), so LOGIN is used as a backup.

kop316 commented:
Hello!

I saw that https://github.com/curl/curl/pull/10077 was added, but I also noticed that all of the tests are in `DISABLED`. I readded the test, but like those, I added to to DISABLED.

I also wanted to circle back on https://github.com/curl/curl/pull/10027#issuecomment-1345580479 . My understanding is still the same as when I did this, so I would like to understand more about it, or if it is an issue?

jay commented:
> I also wanted to circle back on [#10027 (comment)](https://github.com/curl/curl/pull/10027#issuecomment-1345580479) . My understanding is still the same as when I did this, so I would like to understand more about it, or if it is an issue?

@monnerat can you help @kop316 address your concerns?

jay commented:
Closing in favor of #10041 which will add support for preferring plaintext LOGIN over SASL LOGIN via login options.
